### PR TITLE
Split headers body

### PR DESCRIPTION
**Pull request by JunTaoLuo**
- [x] Add an implementation of FastGuid
- [x] Use Guid based key generation
- [x] Update logic for normalizing and comparing cachedVaryBy entries 
- [x] Add an option to configure the size at which head and body are stored separately

cc @Tratcher @davidfowl

### CODE REVIEW

File: src/Microsoft.AspNetCore.ResponseCaching/Interfaces/IKeyProvider.cs [hunk 1]
```diff
@@ -8,18 +8,25 @@ namespace Microsoft.AspNetCore.ResponseCaching
...
-        /// <returns>The created base key.</returns>
+        /// <returns>The created vary key.</returns>
         string CreateVaryKey(HttpContext httpContext, VaryRules varyRules);
+
+        /// <summary>
+        /// Create a key for the response body using the HTTP context.
+        /// </summary>
+        /// <param name="httpContext">The <see cref="HttpContext"/>.</param>
+        /// <returns>The created body key.</returns>
+        string CreateBodyKey(HttpContext httpContext);
```
Comment (JunTaoLuo) [inline src/Microsoft.AspNetCore.ResponseCaching/Interfaces/IKeyProvider.cs]: Though this is sufficient for us, it doesn't really feel right. The logic should be: given an HttpContext and a CachedResponse without a body, create a body key. I'm also not keen on exposing our internal storage representation of a CachedResponse.

File: src/Microsoft.AspNetCore.ResponseCaching/KeyProvider.cs [hunk 2]
```diff
@@ -153,5 +151,28 @@ public virtual string CreateVaryKey(HttpContext httpContext, VaryRules varyRules
                 _builderPool.Return(builder);
             }
         }
+
+        public virtual string CreateBodyKey(HttpContext httpContext)
```
Comment (JunTaoLuo) [inline src/Microsoft.AspNetCore.ResponseCaching/KeyProvider.cs]: Looking into sharding the body in a separate branch. The logic here may eventually take in a chunk number as well to generate the key for a specific shard.

File: src/Microsoft.AspNetCore.ResponseCaching/ResponseCachingContext.cs [hunk 3]
```diff
@@ -181,19 +190,31 @@ internal void FinalizeCachingHeaders()
...
                 if (!StringValues.IsNullOrEmpty(varyHeaderValue) || !StringValues.IsNullOrEmpty(varyParamsValue))
                 {
-                    var cachedVaryRules = new CachedVaryRules
+                    // Normalize order and casing of vary by rules
+                    NormalizeStringValues(ref varyHeaderValue);
+                    NormalizeStringValues(ref varyParamsValue);
+
+                    // Update vary rules if they are different
+                    if (State.CachedVaryRules == null ||
+                        !StringValues.Equals(State.CachedVaryRules.VaryRules.Params, varyParamsValue) ||
```
Comment (JunTaoLuo) [inline src/Microsoft.AspNetCore.ResponseCaching/ResponseCachingContext.cs]: These comparisons assume that the stored vary rules are normalized. Assuming that we are the only ones creating these cached entries, this assumption holds.

File: src/Microsoft.AspNetCore.ResponseCaching/ResponseCachingOptions.cs [hunk 4]
```diff
@@ -19,6 +19,11 @@ public class ResponseCachingOptions
         public bool CaseSensitivePaths { get; set; } = false;
 
         /// <summary>
+        /// The smallest size in bytes for which the headers and body of the response will be stored separately. The default is set to 1 KB.
+        /// </summary>
+        public long MinimumSplitBodySize { get; set; } = 1024;
```
Comment (JunTaoLuo) [inline src/Microsoft.AspNetCore.ResponseCaching/ResponseCachingOptions.cs]: Suggestions for defaults?

**Comment by JunTaoLuo** [comment]
ping

File: src/Microsoft.AspNetCore.ResponseCaching/Internal/FastGuid.cs [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+    internal class FastGuid
+    {
+        // Base32 encoding - in ascii sort order for easy text based sorting
+        private static readonly string _encode32Chars = "0123456789ABCDEFGHIJKLMNOPQRSTUV";
+        // Global ID
+        private static long GlobalID;
+
+        // Instance components
+        private string _idString;
+        internal long ID { get; private set; }
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.ResponseCaching/Internal/FastGuid.cs]: Value?

File: src/Microsoft.AspNetCore.ResponseCaching/Internal/FastGuid.cs [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+using System.Threading;
+
+namespace Microsoft.AspNetCore.ResponseCaching.Internal
+{
+    internal class FastGuid
+    {
+        // Base32 encoding - in ascii sort order for easy text based sorting
+        private static readonly string _encode32Chars = "0123456789ABCDEFGHIJKLMNOPQRSTUV";
+        // Global ID
+        private static long GlobalID;
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.ResponseCaching/Internal/FastGuid.cs]: NextValue?

File: src/Microsoft.AspNetCore.ResponseCaching/ResponseCachingContext.cs [hunk 7]
```diff
@@ -227,9 +249,25 @@ internal void FinalizeCachingBody()
         {
             if (State.ShouldCacheResponse && ResponseCacheStream.BufferingEnabled)
             {
-                State.CachedResponse.Body = ResponseCacheStream.BufferedStream.ToArray();
+                if (ResponseCacheStream.BufferedStream.Length >= _options.MinimumSplitBodySize)
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.ResponseCaching/ResponseCachingContext.cs]: This doesn't give you precise control of your cache entry size because it does not account for the size of the rest of the data (e.g. Headers).

File: src/Microsoft.AspNetCore.ResponseCaching/ResponseCachingContext.cs [hunk 8]
```diff
@@ -227,9 +249,25 @@ internal void FinalizeCachingBody()
...
+                        Body = ResponseCacheStream.BufferedStream.ToArray()
+                    };
 
-                _cache.Set(State.VaryKey ?? State.BaseKey, State.CachedResponse, State.CachedResponseValidFor);
+                    _cache.Set(_keyProvider.CreateBodyKey(_httpContext), cachedResponseBody, State.CachedResponseValidFor);
+                }
+                else
+                {
+                    State.CachedResponse.ContainsBody = true;
+                    State.CachedResponse.Body = ResponseCacheStream.BufferedStream.ToArray();
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.ResponseCaching/ResponseCachingContext.cs]: Do you need ContainsBody or could you just use Body != null?

File: src/Microsoft.AspNetCore.ResponseCaching/ResponseCachingContext.cs [hunk 9]
```diff
@@ -275,5 +313,20 @@ internal void UnshimResponseStream()
             // TODO: Move this temporary interface with endpoint to HttpAbstractions
             _httpContext.RemoveResponseCachingFeature();
         }
+
+        // Normalize order and casing
+        internal static void NormalizeStringValues(ref StringValues stringVales)
+        {
+            var stringArray = stringVales.ToArray();
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.ResponseCaching/ResponseCachingContext.cs]: ToArray is not required, StringValues has an indexer and Count

File: src/Microsoft.AspNetCore.ResponseCaching/KeyProvider.cs [hunk 10]
```diff
@@ -153,5 +151,28 @@ public virtual string CreateVaryKey(HttpContext httpContext, VaryRules varyRules
...
+            {
+                throw new ArgumentNullException(nameof(httpContext));
+            }
+
+            var builder = _builderPool.Get();
+
+            try
+            {
+                return builder
+                    .Append(httpContext.GetResponseCachingState().CachedResponse.Guid.IdString)
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.ResponseCaching/KeyProvider.cs]: It's probably more efficient to use `string.Concat` here since there's only 3 parts.

File: src/Microsoft.AspNetCore.ResponseCaching/ResponseCachingContext.cs (hunk 9 above)
Comment (Tratcher) [inline src/Microsoft.AspNetCore.ResponseCaching/ResponseCachingContext.cs]: Don't update a user's StringValues in place, clone it.

File: src/Microsoft.AspNetCore.ResponseCaching/ResponseCachingOptions.cs (hunk 4 above)
Comment (Tratcher) [inline src/Microsoft.AspNetCore.ResponseCaching/ResponseCachingOptions.cs]: 8kb?

File: src/Microsoft.AspNetCore.ResponseCaching/ResponseCachingContext.cs (hunk 7 above)
Comment (JunTaoLuo) [inline src/Microsoft.AspNetCore.ResponseCaching/ResponseCachingContext.cs]: Offline discussion, we decided that the accuracy here is not significant. Header sizes are relatively small and quite consistent.

File: src/Microsoft.AspNetCore.ResponseCaching/ResponseCachingContext.cs [hunk 11]
```diff
@@ -103,16 +103,8 @@ internal ResponseCachingState State
...
-                        {
-                            body = State.CachedResponse.Body;
-                        }
-                        else
-                        {
-                            var cachedResponseBody = _cache.Get(_keyProvider.CreateBodyKey(_httpContext)) as CachedResponseBody;
-                            body = cachedResponseBody.Body;
-                        }
+                        var body = State.CachedResponse.Body ??
+                            ((CachedResponseBody)_cache.Get(_keyProvider.CreateBodyKey(_httpContext))).Body;
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.ResponseCaching/ResponseCachingContext.cs]: You need to account for the body entry being missing.

File: src/Microsoft.AspNetCore.ResponseCaching/CacheEntry/CacheEntrySerializer.cs [hunk 12]
```diff
@@ -157,8 +186,16 @@ public static void Write(BinaryWriter writer, object entry)
...
+        private static void WriteCachedResponseBody(BinaryWriter writer, CachedResponseBody entry)
+        {
+            writer.Write(entry.Body.Length);
+            writer.Write(entry.Body);
+        }
+
+        // See serialization format above
         private static void WriteCachedResponse(BinaryWriter writer, CachedResponse entry)
         {
+            writer.Write(entry.Guid.IdValue);
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.ResponseCaching/CacheEntry/CacheEntrySerializer.cs]: Reminder: this value is only for responses with bodies cached separately.

File: src/Microsoft.AspNetCore.ResponseCaching/CacheEntry/CachedResponse.cs [hunk 13]
```diff
@@ -8,6 +8,8 @@ namespace Microsoft.AspNetCore.ResponseCaching.Internal
 {
     internal class CachedResponse
     {
+        internal FastGuid Guid { get; set; }
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.ResponseCaching/CacheEntry/CachedResponse.cs]: BodyId

File: src/Microsoft.AspNetCore.ResponseCaching/CacheEntry/CachedVaryRules.cs [hunk 14]
```diff
@@ -0,0 +1,12 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+namespace Microsoft.AspNetCore.ResponseCaching.Internal
+{
+    internal class CachedVaryRules
+    {
+        internal FastGuid Guid { get; set; }
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.ResponseCaching/CacheEntry/CachedVaryRules.cs]: KeyPrefix. This could even be a string, at this point you don't care that it was derived from a FastGuid.

File: src/Microsoft.AspNetCore.ResponseCaching/CacheEntry/CachedResponse.cs (hunk 13 above)
Comment (Tratcher) [inline src/Microsoft.AspNetCore.ResponseCaching/CacheEntry/CachedResponse.cs]: See CachedVaryRules.Guid

File: src/Microsoft.AspNetCore.ResponseCaching/CacheEntry/CachedVaryRules.cs (hunk 14 above)
Comment (JunTaoLuo) [inline src/Microsoft.AspNetCore.ResponseCaching/CacheEntry/CachedVaryRules.cs]: VaryKeyPrefix?

File: src/Microsoft.AspNetCore.ResponseCaching/CacheEntry/CachedResponse.cs (hunk 13 above)
Comment (JunTaoLuo) [inline src/Microsoft.AspNetCore.ResponseCaching/CacheEntry/CachedResponse.cs]: BodyKeyPrefix?

File: src/Microsoft.AspNetCore.ResponseCaching/CacheEntry/CachedVaryRules.cs [hunk 15]
```diff
@@ -0,0 +1,12 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+namespace Microsoft.AspNetCore.ResponseCaching.Internal
+{
+    internal class CachedVaryRules
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.ResponseCaching/CacheEntry/CachedVaryRules.cs]: These types will need to be public to make extensibility possible.

File: src/Microsoft.AspNetCore.ResponseCaching/CacheEntry/CacheEntrySerializer.cs (hunk 12 above)
Comment (JunTaoLuo) [inline src/Microsoft.AspNetCore.ResponseCaching/CacheEntry/CacheEntrySerializer.cs]: Yea this will be reworked.

File: src/Microsoft.AspNetCore.ResponseCaching/CacheEntry/CachedVaryRules.cs (hunk 15 above)
Comment (JunTaoLuo) [inline src/Microsoft.AspNetCore.ResponseCaching/CacheEntry/CachedVaryRules.cs]: Noted. Will do so in a different PR.

File: src/Microsoft.AspNetCore.ResponseCaching/KeyProvider.cs (hunk 2 above)
Comment (Tratcher) [inline src/Microsoft.AspNetCore.ResponseCaching/KeyProvider.cs]: I think you can just remove this method, we don't expect it to exist in the next itteration

Comment (Tratcher) [inline src/Microsoft.AspNetCore.ResponseCaching/KeyProvider.cs]: And the implementation is accessing magic internal state

File: src/Microsoft.AspNetCore.ResponseCaching/ResponseCachingContext.cs (hunk 7 above)
Comment (Tratcher) [inline src/Microsoft.AspNetCore.ResponseCaching/ResponseCachingContext.cs]: Offline discussion: Move the sharding to an implementation detail of the IResponseCache. E.g. there's no reason to shard for memory cache.

File: src/Microsoft.AspNetCore.ResponseCaching/ResponseCachingOptions.cs [hunk 16]
```diff
@@ -19,6 +19,11 @@ public class ResponseCachingOptions
         public bool CaseSensitivePaths { get; set; } = false;
 
         /// <summary>
+        /// The smallest size in bytes for which the headers and body of the response will be stored separately. The default is set to 70 KB.
+        /// </summary>
+        public long MinimumSplitBodySize { get; set; } = 70 * 1024;
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.ResponseCaching/ResponseCachingOptions.cs]: I anticipate this will move to a separate DistributedResponseCacheOptions

File: src/Microsoft.AspNetCore.ResponseCaching/CacheEntry/CachedResponse.cs (hunk 13 above)
Comment (Tratcher) [inline src/Microsoft.AspNetCore.ResponseCaching/CacheEntry/CachedResponse.cs]: sure

File: src/Microsoft.AspNetCore.ResponseCaching/CacheEntry/CachedVaryRules.cs (hunk 14 above)
Comment (Tratcher) [inline src/Microsoft.AspNetCore.ResponseCaching/CacheEntry/CachedVaryRules.cs]: Sure

**Comment by Tratcher** [comment]
:shipit: after removing CreateBodyKey

File: src/Microsoft.AspNetCore.ResponseCaching/ResponseCachingOptions.cs (hunk 16 above)
Comment (JunTaoLuo) [inline src/Microsoft.AspNetCore.ResponseCaching/ResponseCachingOptions.cs]: Yea
